### PR TITLE
#93a CP incentives for healing players

### PR DESCRIPTION
**Pull request by spooker0**
Added:
Functionality to reward CP for healing players

Parameter additions:
```
#define WL_HEAL_SCORE 20
#define WL_HEAL_SCORE_FREQUENCY 20
#define WL_HEAL_SCORE_AISELFHEAL false
```
`WL_HEAL_SCORE`: CP reward per player healed
`WL_HEAL_SCORE_FREQUENCY`: frequency (in seconds) a player can be rewarded for healing
`WL_HEAL_SCORE_AISELFHEAL`: give points to players if their own AI heals them, disabled by default

### CODE REVIEW

File: Functions/client/fn_WL2_initClient.sqf [hunk 1]
```diff
@@ -246,6 +246,11 @@ player addEventHandler ["HandleDamage", {
 	};
 }];
 
+player addEventHandler ["HandleHeal", {
+	params ["_injured", "_healer"];
+	[_injured, _healer] remoteExec ["BIS_fnc_WL2_healRewardHandle", 2];
```
Comment (spooker0) [inline Functions/client/fn_WL2_initClient.sqf:251]: Offload all reward logic to server to possibly improve security (?), open to changes here.

File: Functions/client/fn_WL2_killRewardClient.sqf [hunk 2]
```diff
@@ -19,20 +19,29 @@ _displayY = safeZoneH + safeZoneY - _displayH - (_blockH * 50); //lower vaule he
...
 _ctrl = (findDisplay 46) ctrlCreate ["RscStructuredText", _ctrlNmbr];
 
 _ctrl ctrlSetPosition [_displayX - (_blockW * 110), _displayY - (_blockH * 30), _blockW * 160, _blockH * 16];
 
-if (_unit isKindOf "Man") then {
-	_ctrl ctrlSetStructuredText parseText format ["<t size='0.9' align='right' color='#228b22'>Enemy killed +%1CP</t>", _reward];
-} else {
-	_displayName = getText (configFile >> "CfgVehicles" >> (typeOf _unit) >> "displayName");
-	_ctrl ctrlSetStructuredText parseText format ["<t size='0.9' align='right' shadow = '1' color='#228b22'>%1 destroyed +%2CP</t>", _displayName, _reward];
+_scoreMessage = "Error: No score message. | File: fn_WL2_killRewardClient.sqf | Line: 29";
```
Comment (spooker0) [inline Functions/client/fn_WL2_killRewardClient.sqf:29]: Using the killRewardClient for this to maintain consistency. Should probably rename this func/file to scoreRewardClient in the future.

File: Functions/server/fn_WL2_healRewardHandle.sqf [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	waitUntil {
+		sleep 1;
+		damage _injured != _damage || (time - _startTime > 8);	// healing should take no longer than 8s
+	};
+	
+	_rewardee = leader _healer;
+
+	_lastTimeHeal = _rewardee getVariable "lastTimeHealPoints";
+	if (isNil "_lastTimeHeal") then { 
+		_lastTimeHeal = 0;
```
Comment (spooker0) [inline Functions/server/fn_WL2_healRewardHandle.sqf:18]: Bug/Feature: you will not be able to get heal points within the first X seconds (by default, 20) of the match. Change this line to `-WL_HEAL_SCORE_FREQUENCY` to eliminate this.

**Comment by spooker0** [comment]
Comments added, ready for review.

**Comment by MrThomasM** [comment]
Most likely won't add this, its very easy for cheaters to rexploit this. and even for players there will be some bugs to get free CP.

**Comment by MrThomasM** [comment]
I'll have a look if maybe we can add this in a secure way.

**Comment by spooker0** [comment]
Fair. 
I did attempt to throttle the rate at which you can get rewarded. Unfortunately, I have no idea what the vector for cheating would look like, but it's possible `lastTimeHealPoints` is something that would be accessible to an attacker. Not sure how to protect that. Another possible measure is we can attach this to a player life and reward them for having healed teammates when they die. 
Anyway, the code is yours to play around with.

**Comment by MrThomasM** [comment]
It's a good idea but not secure enough to add and will come with to much exploits for cheaters and players.

**Pull request closed without merging**